text/0001-poet2-consensus.md

```diff
@@ -0,0 +1,161 @@
+* WaitTime: The number of seconds to wait, determined by the LocalMean
+* BlockID: The BlockID passed in to the Enclave
+* PrevBlockID: The BlockID of the previous block, as stored in the previousWaitCertificate
+* TxnHash: The hash of the transactions in the block
+* BlockNumber: The length of the chain
+* ValidatorID: The ID of the current Validator
+* Sign: The signature of the WaitCertificate computed over all the fields using the PSK 
+
+Once the WaitCertificate is created, the enclave stores the Last Block # and the WaitCertificate in the table created previously. The validator will only create a single wait certificate per block. This implies that if a fork containing fewer blocks becomes active, the validator will NOT be able to publish any more blocks until the new fork has added enough blocks to catch up to the previous chain.
+
+## Block Publishing
+With the Claim Block ready and the WaitCertificate created, the validator will determine the duration for which to set the timer. The duration of the wait is determined by the ‘Duration’ field in the WaitCertificate.
+Ideally, any selected algorithm should be capable of converting the duration to a wait time distributed uniformly between 0 and LocalMean seconds.
+
+## Block Verification
+Upon receiving a Claim Block and a WaitCertificate, the receiving validator does the following:
+### Sanity checks
+1.	Ensure the Block and WaitCertificate signatures are valid
+2.	The Block’s ID matches the WaitCertificate’s blockID
+3.	If the block extends the current chain head, proceed to the Block Validity check. If it is a ‘future’ block, cache it for processing later.
+4.	If the block claims a previous block as the parent, proceed to ‘Fork Resolution’
+### Validity check & block publishing
+5.	Compute the new CC. The block is valid if CC + E < WC. Here E is a network latency coefficient added to simulate the average network delay seen by the validator.
+6.	Send the block to the validator for processing the transactions
+7.	If valid, commit the block, store the blockID and WaitCertificate in SGX
+8.	Forward the block to the gossip network
+9.	Drop the existing block proposal and restart from the new chain head.
+### Early Arriving blocks
+10.	If the block is an early-arriving block, compare the duration with the duration of the Claim Block. If the block duration is smaller, drop the Claim Block. Wait for the WC to catch up with the CC. If the Claim Block duration is less than the incoming block duration, discard the incoming block.
+11.	Proceed to step 6 once the block becomes valid.
+### Fork Resolution
+12.	If the new block claims an earlier block as a parent, check for any cached blocks claiming the new block as a parent, retrieve them. Compute the new CC and check if the chain is valid (CC + E < WC)
+13.	If valid, compare existing chain length with the new chain length. If existing chain length is more, discard new chain. If the new chain is longer, switch to the new chain (commit the new blocks)
```

Comment:
I have a question about this fork resolution process.  Can someone explain why we only need to compare chain lengths (rather than check the total number of "work" done on the chain--i.e., incorporate the population estimate)?

Here's my thought process:  suppose I controlled a small group of nodes.  We could could fork off everything to a side chain and then wait for the "difficulty" to go down.  In the steady state, due to the population adjustment, we would be adding blocks at the same rate as the main chain.  We could let this go on for a while--maybe we would get lucky and accumulate blocks at a slightly (sublinear) rate faster than the main chain.  Then, at some point, we could add in a bunch of new members and use these to add much more blocks than expected and try to catch up to the main chain.  It's possible there's some mathematical reasoning that says this is impossible, but it's not immediate (and needs to be written up if this is becomes the spec).

Does anyone have an explanation for why this doesn't work?  Am I misunderstanding something here?  Thanks!

Reply:
@hartm  The newer spec for poet 2.0 is at https://github.com/hyperledger/sawtooth-rfcs/pull/20